src/main/kotlin/com/manimdsl/ASTExecutor.kt

```diff
@@ -14,35 +14,102 @@ data class StackValue(val initObject: MObject, val stack: Stack<Pair<Double, MOb
 
     private val linearRepresentation = mutableListOf<ManimInstr>()
     private val variableNameGenerator = VariableNameGenerator(symbolTableVisitor)
-
-    private val codeBlockVariable: String
-    private val codeTextVariable: String
-    private val pointerVariable: String
+    private val codeBlockVariable: String = variableNameGenerator.generateNameFromPrefix("code_block")
+    private val codeTextVariable: String = variableNameGenerator.generateNameFromPrefix("code_text")
+    private val pointerVariable: String = variableNameGenerator.generateNameFromPrefix("pointer")
+    private val displayLine: MutableList<Int> = mutableListOf()
+    private val displayCode: MutableList<String> = mutableListOf()
 
     init {
-        pointerVariable = variableNameGenerator.generateNameFromPrefix("pointer")
-        codeBlockVariable = variableNameGenerator.generateNameFromPrefix("code_block")
-        codeTextVariable = variableNameGenerator.generateNameFromPrefix("code_text")
+        fileLines.indices.forEach {
+            val presentableLine = fileLines[it] == "}" || fileLines[it] == "{" || fileLines[it] == "" || (statements.containsKey(it + 1) && statements[it + 1] is CodeNode)
+            if (presentableLine) {
+                displayCode.add(fileLines[it])
+                displayLine.add(1 + (displayLine.lastOrNull() ?: 0))
+            } else {
+                displayLine.add(displayLine.lastOrNull() ?: 0)
+            }
+        }
     }
 
-    // Map of all the variables and there current values
-    private val variables = mutableMapOf<String, ExecValue>()
+    fun runProgram(): List<ManimInstr> {
+        linearRepresentation.add(CodeBlock(displayCode, codeBlockVariable, codeTextVariable, pointerVariable))
+        val variables = mutableMapOf<String, ExecValue>()
+        Frame(program.statements.first().lineNumber, fileLines.size, variables).runFrame()
+        return linearRepresentation
+    }
+
+    private inner class Frame(private var pc: Int, private var finalLine: Int, private var variables: MutableMap<String, ExecValue>) {
+
+        // instantiate new Frame and execute on scoping changes e.g. recursion
+
+        fun runFrame(): ExecValue {
 
-    // Keep track of next instruction to execute
-    private var programCounter = 0
+            while (pc <= finalLine) {
+
+                if (statements.containsKey(pc)) {
+                    val statement = statements[pc]
+
+                    if (statement is CodeNode) {
+                        moveToLine()
+                    }
+
+                    when (statement) {
+                        is ReturnNode -> {
+                            return executeExpression(statement.expression)
+                        }
+                        is FunctionNode -> {
+                            // just go onto next line, this is just a label
```

Comment:
Can you remove this case if we are skipping it? Or put like a continue for clarity